[2단계 - 출석 다시 구현하기] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
1단계와는 다르게 도메인 객체들이 분명한 책임을 가지고 협력하는 구조로 설계했다고 생각합니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

안녕하세요 제이, 2단계 미션 제출합니다.😀
리뷰 잘부탁드려요!

이번에는 미션 목적에 걸맞게 많은 부분에서 TDD 방식을 적용해보았어요.

TDD 방식으로 개발을 시작하기 전에 어떤 객체들이 어떤 책임을 가질 것인가 미리 생각한 다음에,
```a객체가 b객체에게 무엇을 넘겨서 무엇을 반환할 것을 요청한다``` 이런 식으로 애플리케이션의 흐름을 쭉 작성해보았습니다.
그리고 테스트하기 쉬운 가장 작은 단위의 기능을 찾아 실패하는 테스트를 먼저 작성하고 해당 책임의 객체를 생성하여 테스트를 성공할 수 있도록 구현했습니다.

일부 기능에서는 TDD 사이클에 맞게 test(red), feat(green)으로 나누어 커밋을 했습니다.

확인해주시고 부족한 점 피드백 주시면 감사하겠습니다.😃

---

추가로 이번 2단계를 진행하며
```언제가 예외를 던져야 할 상황일까?``` 에 대해 고민을 많이 했어요.

습관적으로 예외를 던지고 있던 상황이 과연 진짜 예외 상황일까하는 의문이 들었습니다.
예를 들어 날짜를 형식에 맞추어 잘 입력했는데, 기능 요구 사항인 캠퍼스 등교일이 아니라는 이유만으로 예외를 던지는 것이 맞을까하는 생각들이에요.

예외를 던지지 않아도 인자로 받은 날짜가 등교일이 맞는 지 true/false를 반환하는 메서드만 있으면.
false를 받았을 때 Application에서 다음 로직을 진행하지 않는 방식으로 구현할 수 있겠다고 생각했거든요. 

외에도 크루 이름이 존재하지 않는 경우 / 캠퍼스 운영 시간이 아닌 경우 / 이미 출석한 경우 등이 예외를 던질만한 상황인가에 대해서도 고민했어요.

우선 저는 예외를 던지지 않아도 구현이 가능하다고 생각했기 때문에 예외를 던지지 않는 방향으로 기능 구현을 했습니다.

다만 아래처럼 모든 케이스를 각각 처리하려니 처리 지점이 많아지고, 처리하는 쪽의 메서드 코드가 점점 많아져 코드의 가독성이 떨어진다고 느꼈어요.
```
private static void doAttendance() {
        LocalDate today = now();
        boolean isOperationDate = CampusManager.isOperationDate(today);
        if (!isOperationDate) {
            OutputView.printNotOperationDate(today);
            return;
        }
        Nickname crewNickname = new Nickname(readCrewNickname());
        boolean isCrewExists = attendanceManager.isCrewExists(crewNickname);
        if (!isCrewExists) {
            OutputView.printNotRegisteredCrew();
            return;
        }
        Optional<Attendance> existingAttendance = attendanceManager.findAttendance(crewNickname, today);
        if (existingAttendance.isPresent()) {
            OutputView.printDuplicatedAttendance();
            return;
        }
        LocalTime attendanceTime = readAttendanceTime();
        boolean isOperationTime = CampusManager.isOperationTime(attendanceTime);
        if (!isOperationTime) {
            OutputView.printNotOperationTime();
            return;
        }
        Attendance attendance = attendanceManager.addAttendance(crewNickname, today, attendanceTime);
        OutputView.printAttendance(attendance);
    }
```

결국 이러한 단점들이 크게 다가와 아래 코드처럼 예외를 던지고 한 곳에서 try catch를 하도록 일부 리팩터링했습니다.

```
    private void doAttendance() {
        validateOperationDate(today);
        Nickname crewNickname = new Nickname(readCrewNickname());
        attendanceManager.validateExistingCrew(crewNickname);
        attendanceManager.validateDuplicatedAttendance(crewNickname, today);
        LocalTime attendanceTime = readAttendanceTime();
        validateOperationTime(attendanceTime);
        Attendance attendance = attendanceManager.addAttendance(crewNickname, today, attendanceTime);
        printAttendance(attendance);
    }
```

이 과정을 통해 제가 느낀 예외를 던졌을 때의 장점은 다음과 같았어요.
- 한 곳에서 처리할 수 있다.
- 개발자의 의도가 명확하게 드러난다.
- 예외를 던지는 것까지를 책임으로 볼 수 있다.

질문이 다소 모호한 것 같지만,
제이는 이러한 고민과 리팩터링 전/후에 대해 어떻게 생각하시는 지 궁금합니다.😀

---

마지막으로 질문이 하나 더 있습니다..!
```
private void doAttendance() {
        validateOperationDate(today);
        Nickname crewNickname = new Nickname(readCrewNickname());
        attendanceManager.validateExistingCrew(crewNickname);
        attendanceManager.validateDuplicatedAttendance(crewNickname, today);
        LocalTime attendanceTime = readAttendanceTime();
        validateOperationTime(attendanceTime);
        Attendance attendance = attendanceManager.addAttendance(crewNickname, today, attendanceTime);
        printAttendance(attendance);
    }
```
저는 위 코드에서 3번의 검증 로직(validate###)을 거쳤기 때문에 ```attendanceManager```의 ```addAttendance``` 내부에서 같은 검증 로직을 거치지 않게끔 구현했습니다.
(앞에 미리 검증 로직을 가졌어야 하는 이유는 기능 요구 사항의 검증 순서 때문입니다.)

그러다가 미래의 나 혹은 제 3자가 검증 로직을 생략하고 ```addAttendance```을 직접 호출한다면 개발자가 의도하지 않은 예외가 발생할 수 있겠다는 생각이 들었어요.

제이는 이런 상황을 마주하시면 보통 어떤 선택을 하시는 지 궁금합니다.
이미 검증 로직을 거쳤기 때문에 내부에서 다시 검증을 하지 않으신가요?
아니면 내부에서도 다시 검증 로직을 거치도록 하나요?

감사합니다!

kjyyjk committed:
chore: step1 코드 전부 삭제

kjyyjk committed:
docs: 기능 목록 작성

kjyyjk committed:
feat: 주어진 날짜가 캠퍼스 등교일인지 여부를 반환

- 캠퍼스 운영과 관련된 책임을 가지는 CampusManager는 넘겨 받은 날짜가
캠퍼스 등교일(주말, 공휴일 제외)인지를 여부를 확인하고 반환할 수 있다.

- 12월 한정 운영이므로 공휴일은 12월 25일 크리스마스만을 고려한다.

kjyyjk committed:
feat: 주어진 시간이 캠퍼스 운영 시간인지 여부를 반환

- 캠퍼스 운영과 관련된 책임을 가지는 CampusManager는 넘겨 받은 시간이
캠퍼스 운영시간(매일 8:00 ~ 23:00)인지 여부를 확인하고 반환할 수 있다.

kjyyjk committed:
chore: attendance 패키지 추가

kjyyjk committed:
feat: 주어진 요일과 시간을 기반으로 출석 상태를 반환

- 출석 정책과 관련된 책임을 가지는 AttendancePolicy는 넘겨 받은 요일과
출석 시간을 기반으로 출석 상태를 반환할 수 있다.

kjyyjk committed:
refactor: Nested 중첩클래스를 활용해 테스트 그룹화

kjyyjk committed:
feat: 닉네임이 같으면 동일한 크루

- 크루의 닉네임은 중복을 허용하지 않는다.
따라서 Crew의 nickname이 같으면 같은 객체라고 판단한다.

kjyyjk committed:
feat: 크루가 존재하는 지 여부 반환

- 크루 별 출석 기록을 관리할 AttendanceManager는
파라미터로 받은 크루가 존재하는 지 여부를 반환할 수 있다.

kjyyjk committed:
feat: 주어진 날짜와 Attendance의 날짜가 같은 지 여부 반환

kjyyjk committed:
feat: 주어진 날짜의 Attendance를 찾아 반환

- AttendanceHistory는 특정 크루의 Attendance들을 보관한다.

- 주어진 날짜와 동일한 Attendance 객체를 찾아 Optional로
감싸 반환한다. 만약 없으면 empty로 반환한다.

kjyyjk committed:
feat: 출석 기능 구현 및 크루의 주어진 날짜 Attendance 반환

- 출석을 관리하는 AttendanceManager에 크루 이름과 출석 날짜, 출석 시간을
넘겨 출석을 저장할 수 있다.

- 출석을 관리하는 AttendanceManager에 크루 이름과 출석 날짜를 넘겨
해당 날짜의 크루 출석을 받아올 수 있다. 만약 출석이 없다면 Optional.empty를
반환한다.

kjyyjk committed:
feat: 출석에 시간 상태 추가

- Attendance가 출석 시간 정보를 가지게끔 LocalTime을 상태로 추가하였다.

kjyyjk committed:
refactor: domain 패키지 분리

- 도메인 객체들을 attendance/domain 패키지에 보관한다.

kjyyjk committed:
feat: 옵션 입력

- 사용자로부터 입력을 받아오는 InputView를 생성한다.

- UI 관련 클래스를 보관하는 view 패키지에 보관한다.

kjyyjk committed:
feat: 입력에 따른 기능 실행 구현

- 옵션을 key로 가지고, Runnable 기능을 value로 가지는
map을 활용한다. 이 덕분에 option별 if 분기문을 작성하지 않아도
옵션 별 기능을 실행할 수 있다.

kjyyjk committed:
feat: 존재하지 않는 옵션일 경우 예외 발생

kjyyjk committed:
feat: 예외 처리 및 옵션 재입력

- 예외가 발생하면 Application에서 예외 메세지를 출력 후,
옵션부터 재입력을 받는다.

kjyyjk committed:
feat: 캠퍼스 등교일이 아니면 안내 후 재입력

- 출석하는 오늘이 캠퍼스 등교일이 아니라면, 안내 메세지를
출력 후에 옵션부터 재입력받는다.

- 예외를 던지지 않고 true/false로 판별한다.

kjyyjk committed:
feat: 출석 시 크루 닉네임 입력

kjyyjk committed:
feat: 존재하지 않는 닉네임일 경우 안내 후 재입력

kjyyjk committed:
feat: 입력한 등교 시간이 유효하지 않을 경우 예외 발생

kjyyjk committed:
feat: 캠퍼스 운영 시간 외 등교 시 안내 후 옵션 재입력

kjyyjk committed:
feat: 이미 출석한 경우 안내하고 옵션부터 재입력

kjyyjk committed:
feat: 출석을 저장

kjyyjk committed:
feat: 출석 결과 출력

kjyyjk committed:
feat: 출석 수정 시 닉네임 입력

kjyyjk committed:
feat: 출석 수정 중 존재하지 않는 크루일 시 처리

kjyyjk committed:
feat: 출석 수정 날짜 입력

kjyyjk committed:
feat: 유효하지 않은 출석 수정 날짜 입력 시 처리

- 숫자가 아닌 문자를 입력한 경우 발생하는 InputMismatchException과
올바르지 않은 날짜를 입력했을 경우 발생하는 DateTimeException을
잡아 IllegalArgumentException으로 던진다.

kjyyjk committed:
rename: 수정 대상 날짜의 변수명 변경

- 기존에는 modifiedAttendanceDate였으나 이는 수정된 출석의 날짜라는
의미로 문맥상 어색하다. 따라서 수정될 출석의 날짜라는 의미의
attendanceDateToModify로 변경한다.

kjyyjk committed:
feat: 수정 날짜가 등교일이 아닌 경우 처리

kjyyjk committed:
fix: 날짜 입력 후 버퍼의 잔여 개행 처리

- nextInt()로 수정 날짜를 입력 받을 때 엔터로 인한 개행이
버퍼에 남아있게 된다. 이로 인해 다음 입력 시 개행이 자동으로 들어가며
의도치 않은 입력처리가 된다. 따라서 nextInt() 이후 nextLine()을
해주어 잔여 개행을 처리한다.

kjyyjk committed:
feat: 출석 수정 시간 입력

kjyyjk committed:
feat: 출석 수정 시간이 캠퍼스 운영 시간이 아닌 경우 처리

kjyyjk committed:
rename: 파라미터에 정보가 주어지므로 메서드 끝의 By구 제거

kjyyjk committed:
feat: 수정할 출석 기록이 없을 경우 처리

kjyyjk committed:
test(red): AttendanceHistory 출석 수정 테스트

- AttendanceHistory는 기존 출석을 인자로 받은 수정된 출석으로 변경할
수 있다.

kjyyjk committed:
feat(green): AttendanceHistory 출석 수정 구현

- 수정 이전의 Attendance를 제거하고 수정된 Attedance를 새로 추가한다.
이후 수정된 Attendance를 반환한다.

kjyyjk committed:
test(red): AttendanceManager 출석 수정 테스트

kjyyjk committed:
feat(red): AttendanceManager 출석 수정 구현

kjyyjk committed:
feat: 출석 수정

kjyyjk committed:
feat: 출석 수정 전 후 결과 출력

kjyyjk committed:
test(red): AttendanceStatistics가 위험 상태 계산

kjyyjk committed:
feat(red): AttendanceStatistics가 위험 상태 계산

kjyyjk committed:
refactor: CrewStatus Enum 상수화

- 기존에는 String 타입의 dangerousStatus로 사용하였으나.
사용되는 곳이 많아 유지 보수의 용이함을 위해 Enum 상수화하였다.
이렇게 할 경우 추후 각 상수에 필드를 부여할 수 있어 출력이나
정렬 시에도 용이하다.

kjyyjk committed:
refactor: AttendanceStatus Enum 상수화

- 기존에는 String 타입으로 사용하였으나, 사용되는 곳이 많고 유지 보수의
용이성을 위해 Enum 상수화하였다.

kjyyjk committed:
refactor: 테스트 패키지 이동

kjyyjk committed:
test(red): 출석 기록들을 기반으로 AttendanceStatistics를 반환

kjyyjk committed:
feat(green): 출석 기록들을 기반으로 AttendanceStatistics를 반환

kjyyjk committed:
feat: 출석 기록 확인 시 크루 닉네임 입력

kjyyjk committed:
feat: 출석 기록 확인 시 존재하지 않는 크루 닉네임 처리

kjyyjk committed:
feat: 크루의 출석 기록 조회

kjyyjk committed:
feat: 크루의 출석 기록 출력

- 출력 시 등교일임에도 출석이 없는 경우에는
결석으로 간주하고 출력한다.

kjyyjk committed:
feat: 출석 통계 및 면담 대상자 출력

kjyyjk committed:
feat: 제적 위험자 확인 및 출력

kjyyjk committed:
feat: 프로그램 시작 시 파일 읽어와 출석 동기화

kjyyjk committed:
feat: 출력 요구사항에 맞게 줄바꿈 처리

kjyyjk committed:
feat: 결석 수 동일 시 닉네임 오름차순 정렬

kjyyjk committed:
test: Attendance의 isYearMonthDate 테스트

kjyyjk committed:
test: 테스트 코드 형식 리팩토링

kjyyjk committed:
refactor: AttendancePolicy 분기 리팩터링

- 기존에는 월요일에 따른 분기, 일반 요일에 따른 분기를 모두 작성하였으나
가독성이 떨어진다. 따라서 메서드를 분리하여 중복을 줄이고 가독성을 늘렸다.

kjyyjk committed:
refactor: 매직넘버 상수화

kjyyjk committed:
style: 메서드 파라미터의 줄바꿈 형식 통일

kjyyjk committed:
refactor: static 메서드만을 가지는 클래스의 생성자 private화

kjyyjk committed:
rename: Crew를 Nickname으로 클래스명 변경

- 현재 기능에서 크루가 상태로 가지는 것은 Nickname 외에 존재하지 않는다.
또한 출석부에서도 크루를 가질 필요 없이 크루를 식별할 수 있는 크루의
닉네임을 가지는 것이 적절하다고 생각된다. 따라서 Crew에서 Nickname으로
클래스명을 변경한다.

- 추후 Crew 객체만의 책임이 생긴다면 Crew를 생성하고,
해당 Crew객체를 관리하는 CrewManager를 생성할 수도 있을 것이다.

kjyyjk committed:
refactor: 읽어온 파일을 파싱하고 출석을 초기화 하는 로직을 클래스화

kjyyjk committed:
refactor: 크루가 존재하지 않을 경우 예외를 던지도록 수정

- 기존에는 AttendanceManager에서 boolean을 반환하고, Application에서
boolean 값을 가지고 처리하는 방식이었다. 하지만 코드가 길어져 가독성이
떨어지고 동일 로직에 대한 처리 지점이 많아져 유지보수에 이점이 없다는
판단을 했기에 수정을 했다.

kjyyjk committed:
refactor: 캠퍼스 운영시간이 아닐 경우 예외를 던지도록 수정

- 마찬가지로 코드가 길어져 가독성이 떨어지고 동일 로직에 대한 처리 지점이
많아져 유지보수에 이점이 없다는 판단을 했기에 수정을 했다.

kjyyjk committed:
refactor: 캠퍼스 등교일이 아닐 경우 예외를 던지도록 수정

- 마찬가지로 코드가 길어져 가독성이 떨어지고 동일 로직에 대한 처리 지점이
많아져 유지보수에 이점이 없다는 판단을 했기에 수정을 했다.
이때 에러메세지는 date에 따라 다르게 포매팅되어야 하고 하나의 try catch에서
잡아 처리하고 싶었기에 커스텀 예외인 NotOperationDateException을 만들었다.

- NotOperationDateException은 date를 인자로 받아 내부에서 에러 메세지를
포매팅하고 상위 예외인 IllegalArgumentException의 생성자를 호출하며 메세지를 넘긴다.

kjyyjk committed:
refactor: 중복 출석을 할 경우 예외를 던지도록 수정

- 마찬가지로 코드가 길어져 가독성이 떨어지고 동일 로직에 대한 처리 지점이
많아져 유지보수에 이점이 없다는 판단을 했기에 수정을 했다.

kjyyjk committed:
refactor: 옵션 별 기능 실행 로직 클래스화

- 기존에는 Application에서 프로그램 초기 설정, 프로그램 흐름 처리,
옵션 입력, 옵션 별 기능 실행을 모두 책임졌다. 하지만 이는 너무 많은
책임이 부여되어있다고 판단하여 Options 클래스를 생성했다.

- Options는 옵션과 기능 수행에 대한 책임을 지닌다.

kjyyjk committed:
refactor: 에러 메세지 Enum 상수화

- 도메인, 뷰 등 여러 곳에서 예외를 던지고 있는데, 그 메세지는 어느정도 UI의
영역이라고 생각했기에 따로 Enum 클래스로 분리하였다.

kjyyjk committed:
test: AttendanceStatistics의 출석 상태 횟수 반환 테스트

kjyyjk committed:
test(red): 닉네임 null 또는 공백 포함 검증

kjyyjk committed:
feat(green): 닉네임 null 또는 공백 포함 검증

kjyyjk committed:
refactor: 닉네임 검증 리팩터링

JunHoPark93 commented on README.md:
```diff
@@ -1,48 +1,65 @@
+- [ ] 크루 및 출석 초기화
+  - [x] 파일을 읽어온다.
+  - [x] 크루와 출석 기록을 동기화한다.
+    - [ ] 크루의 닉네임이 공백 또는 null일 경우 예외를 발생한다.
```
완성된 기능이면 체크해주세요~

JunHoPark93 commented on src/main/java/attendance/Options.java:
```diff
@@ -0,0 +1,104 @@
+
+    private Map<String, Runnable> options;
+    private final AttendanceManager attendanceManager;
+    private final LocalDate today;
```
options의 역할에 대해 다시 고민해보면 어떨까요? 출석관리의 역할이 이 클래스의 멤버로 들어가는게 적절할까요?

kjyyjk replied:
제가 생각하는 ```Options```의 역할은 옵션-기능 매핑 정보를 보관하고, 옵션에 따른 기능을 호출하는 것까지 입니다.
제이가 말해주신 것처럼 현재의 ```Options```는 그 책임과 역할을 많이 벗어났다고 생각해요.

kjyyjk replied:
우선 Options 내부에 있던 로직들을 다시 Application으로 옮기고,
Options가 이를 호출하는 쪽으로 수정하였습니다.
bfa0c7a3943f84261dce8c47ff6baff8e220bf0f

JunHoPark93 replied:
options내부가 많이 정리됐네요!

JunHoPark93 commented on src/main/java/attendance/Options.java:
```diff
@@ -0,0 +1,104 @@
+        function.run();
+    }
+
+    private void doAttendance() {
+        validateOperationDate(today);
+        Nickname crewNickname = new Nickname(readCrewNickname());
+        attendanceManager.validateExistingCrew(crewNickname);
+        attendanceManager.validateDuplicatedAttendance(crewNickname, today);
+        LocalTime attendanceTime = readAttendanceTime();
+        validateOperationTime(attendanceTime);
+        Attendance attendance = attendanceManager.addAttendance(crewNickname, today, attendanceTime);
+        printAttendance(attendance);
+    }
```
options클래스의 크기가 너무 큽니다.

더불어 validateOperationTime 등등의 검증의 역할도 options의 역할이 맞을까요?

JunHoPark93 replied:
> 그러다가 미래의 나 혹은 제 3자가 검증 로직을 생략하고 addAttendance을 직접 호출한다면 개발자가 의도하지 않은 예외가 발생할 수 있겠다는 생각이 들었어요.
> 
> 제이는 이런 상황을 마주하시면 보통 어떤 선택을 하시는 지 궁금합니다.
> 이미 검증 로직을 거쳤기 때문에 내부에서 다시 검증을 하지 않으신가요?
> 아니면 내부에서도 다시 검증 로직을 거치도록 하나요?

이부분도 위와 관련이있을거같은데요, 도메인 핵심부의 검증을 제일 중요하게 생각해주셔야 될것같습니다. 역할과 책임을 나누어서 readme에 정리를 한번 하고 진행하면 조금 덜 헷갈리실수도있으니 참고부탁드려요~

아래내용을 고려해주시면 좋을것같습니다.
- attendance를 추가하는 역할은 누가하는것인지
- attendance를 추가할때 어떤 검증을 해야하는지 (코드를 봤을때 attendanceTime을 검증해야하고 이 검증은 추가하는 역할을 가진 도메인내부에서 검증하면 호출부는 검증을 신경쓰지않고 호출할 수 있게 되겠죠)

추가로 말씀주신 상황은 의도치않은 버그가 생길수가 있기때문에 지양해야되는 구조로 생각됩니다.

kjyyjk replied:
조금 더 생각해봤는데 제이의 의견에 동의합니다!
호출부보다 도메인 핵심부의 검증을 가장 중요시하게 여겨야 할 것 같아요.

추가로 기능 순서 상 도메인 핵심부보다 호출부에서 먼저 검증을 필요로 하는 경우가 있기에,
해당 경우에는 둘 다 가능하도록 수정했습니다.

JunHoPark93 commented on src/main/java/attendance/Options.java:
```diff
@@ -0,0 +1,104 @@
+        LocalDate dateToModify = readAttendanceDateToModify(today);
+        validateOperationDate(dateToModify);
+        Optional<Attendance> existingAttendance = attendanceManager.findAttendance(crewNickname, dateToModify);
+        if (existingAttendance.isEmpty()) {
```
> 예외를 던지지 않아도 인자로 받은 날짜가 등교일이 맞는 지 true/false를 반환하는 메서드만 있으면.
> false를 받았을 때 Application에서 다음 로직을 진행하지 않는 방식으로 구현할 수 있겠다고 생각했거든요.
> 
> 외에도 크루 이름이 존재하지 않는 경우 / 캠퍼스 운영 시간이 아닌 경우 / 이미 출석한 경우 등이 예외를 던질만한 상황인가에 대해서도 고민했어요.
> 
> 우선 저는 예외를 던지지 않아도 구현이 가능하다고 생각했기 때문에 예외를 던지지 않는 방향으로 기능 구현을 했습니다.
> 
> 다만 아래처럼 모든 케이스를 각각 처리하려니 처리 지점이 많아지고, 처리하는 쪽의 메서드 코드가 점점 많아져 코드의 가독성이 떨어진다고 느꼈어요.
> 

정의해주신 도메인의 역할에 따라 다를거같은데요, 저는 말씀주신케이스들에 대해서 예외를 발생시켜도 괜찮다고 생각해요.
예외를 공통화 시키고 적절히 처리하는것도 방법이니 고려해보시면 좋을거같네요!

JunHoPark93 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,37 @@
+import java.time.LocalTime;
+
+public class Attendance {
+    private final LocalDate date;
+    private final LocalTime time;
```
date와 time을 나누어 관리하는 이유가 있으신가요?

kjyyjk replied:
Attendance 내부에 날짜 정보를 비교하는 ```isDateEquals()```가 존재하기 때문에, 날짜와 시간을 나누어 관리하면 사소하지만 ```toLocalDate()```와 같은 별도의 변환 로직을 거치지 않을 수 있겠다고 생각하였습니다.

kjyyjk replied:
사실 LocalDateTime으로 관리하는 것 또한 고려해보았었는데, 두 방식 중 어떤 것으로 해도 큰 차이가 없지 않을까하는 마음에 그때 가장 편리했던 방식을 택했던 것 같아요.

kjyyjk replied:
이러한 경우에는 날짜 시간 정보를 하나의 상태로 보관하는 게 더 나은 방법일까요?

JunHoPark93 replied:
인스턴스변수 개수도 가능한한 적게 유지하는게 유지보수 관점에서 좋다고생각해요. 특별히 나눠서 관리해야되는 이유가 있는게 아니라면요!

JunHoPark93 commented on src/main/java/attendance/domain/Nickname.java:
```diff
@@ -0,0 +1,59 @@
+
+import java.util.Objects;
+
+public class Nickname implements Comparable {
```
```suggestion
public class Nickname implements Comparable<NickName> {
```

kjyyjk replied:
Comparable 구현 시 제네릭 타입을 명시해주니 compareTo 메서드 내에서 형 변환을 할 필요가 없네요..!
수정했습니다.

JunHoPark93 commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,85 @@
+        return attendanceHistory.getAttendanceStatistics(today);
+    }
+
+    public Map<Nickname, AttendanceStatistics> getDangerousCrewsInformation(final LocalDate today) {
```
메서드의 depth를 줄이려면 어떻게하면 좋을까요?

kjyyjk replied:
메서드를 분리하여 depth를 감소시켰습니다!

kjyyjk replied:
이 부분에서 제이에게 궁금한게 있는데, 혹시 실무에서도 미션의 프로그래밍 요구 사항처럼 메서드 내 들여쓰기 단계를 한 단계로 유지하려는 경우가 많은가요?

JunHoPark93 replied:
무조건 1단계로 유지하지는 않지만 최대한 단계를 낮게 가져가려고하고있어요.
몇 단계인지가 중요하다기보다 어떤게 더 읽기 편한지를 고려하는 편입니다.

첨언하자면,
depth를 신경쓰지 않게되면 가독성을 해치거나 불필요하게 깊어지는 경우가 생기게 되더라고요. 미션요구사항인 1depth를 지켜보면서 연습을 하게되면 나중에 어떨때는 depth를 더 가져가도 괜찮겠구나에 대한 본인만의 기준이 생길겁니다.

극단적 연습을 통하게되면 추후 유연성이라는 측면을 고려하실때 좀 더 수월하시지 않을까합니다. 
운동선수들이 모래주머니를 차고 훈련을 하지만 실제경기에서는 주머니를 차지않고 뛰는것처럼요.

kjyyjk replied:
한번에 와닿는 비유네요..! 좋은 이야기 감사합니다.

JunHoPark93 commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,85 @@
+
+    public void validateDuplicatedAttendance(final Nickname crewNickname, final LocalDate attendanceDate) {
+        Optional<Attendance> attendance = findAttendance(crewNickname, attendanceDate);
+        if (attendance.isPresent()) {
```
chaining등으로 개선해볼 수 있을까요?

kjyyjk replied:
Optional에서 제공하는 ```ifPresent()```를 활용해서 존재하면 예외를 던지도록 수정했습니다.
추가로 ```addCrew()``` 메서드도 Map에서 제공하는 ```putIfAbsent()```를 활용하도록 내부 구현 수정했습니다.

코드가 더 깔끔해진 것 같아 좋은 거 같아요.

JunHoPark93 commented on src/main/java/attendance/domain/CampusManager.java:
```diff
@@ -0,0 +1,50 @@
+import java.time.LocalDate;
+import java.time.LocalTime;
+
+public class CampusManager {
```
출석관리자와 캠퍼스 관리자를 나눠두신 의도를 말씀주실 수 있을까요?

가능하다면 각각의 도메인의 역할에 대해서 readme에 정리해주시면 좋을거같아요.

kjyyjk replied:
캠퍼스 운영일 혹은 운영 시간을 관리하는 것은 출석을 관리하는 것과는 별개의 기능이라고 보았기 때문에,
책임을 분리하려고 ```CampusManager```와 ```AttendanceManager```를 나누었습니다.

JunHoPark93 commented on src/main/java/attendance/domain/AttendanceHistory.java:
```diff
@@ -1,49 +1,69 @@
-        return attendanceType;
+    public Attendance modifyAttendance(final Attendance modifiedAttendance) {
+        LocalDate attendanceDateToModify = modifiedAttendance.getDate();
+        Attendance beforeAttendance = findAttendance(attendanceDateToModify).get();
```
요부분에서 출석이 없을때는 문제가 발생하지 않을까요?

kjyyjk replied:
수정 뿐만 아니라 출석 추가 등 도메인 핵심부에 검증을 추가했습니다!

JunHoPark93 reviewed (changes requested):
안녕하세요 레오!
2단계 구현하느라 고생많으셨습니다.
질문주신부분들 코멘트남겨두었는데 수정해보시고 어려운점 있으시면 다시 말씀부탁드립니다~

kjyyjk committed:
docs: README 기능 목록 최신화

kjyyjk committed:
refactor: 자바의 기본 메서드들을 활용하여 분기문 리팩토링

- 기존에는 특정 요소가 있는 지를 직접 확인하고, 그 결과를 분기로하여
특정 동작을 수행했다. 하지만 자바 제공 기본 메서드들을 활용하면
더욱 간결하게 작성할 수 있다.

- putIfAbsent는 Map에 주어진 key에 대한 매핑이 존재하지 않으면
 해당 key-value 매핑을 추가한다.

 - ifPresent는 함수형 인터페이스인 Consumer를 파라미터로 받아,
 Optional의 값이 존재하면 Consumer를 실행한다.

kjyyjk committed:
refactor: Comparable 구현 시 제네릭 타입 명시

kjyyjk committed:
feat: 도메인 핵심부 검증 추가

- 기존에는 기능 순서 상 도메인 로직을 호출하기 이전에 호출부 쪽에서
먼저 검증을 거쳤다. 그래서 도메인 로직의 핵심부에서는 따로 검증 로직을
추가할 필요가 없다고 판단하였다. 하지만 제 3자가 검증 로직 호출 없이
핵심부를 호출하면 의도하지 않은 예외가 발생할 수 있다. 그렇기 때문에
도메인 핵심부에 검증을 추가한다.

kjyyjk committed:
refactor: Options의 도메인 로직 Application으로 이동

- 기존에는 Options가 일종의 Controller 역할을 하였으나,
이는 Options의 책임이 아니라고 판단하여 Application으로 이동하였다.
따라서 Options는 입력 받은 Option에 따라 Application의 메서드를
호출한다.

kjyyjk committed:
refactor: main()의 depth 감소

- 기존에는 main 내부에서 종료 옵션인지를 확인하고, 종료 옵션이면
break를 걸어 프로그램을 종료했다.

- Options의 run()이 boolean을 반환하도록 수정하여
계속할 경우 true를, 종료 옵션일 시 false를 반환하도록 한다.
main에서는 again이라는 flag를 활용해 run이 false를 반환하면
반복하지 않도록 한다.

kjyyjk committed:
refactor: 메서드 분리하여 depth 감소

kjyyjk committed:
docs: 클래스 별 역할 정리

JunHoPark93 commented on src/main/java/attendance/domain/AttendanceHistory.java:
```diff
@@ -1,49 +1,82 @@
+    }
+
+    private void addAttendanceStatusCount(final LocalDate date, final Map<AttendanceStatus, Integer> statistics) {
+        boolean isOperationDate = CampusManager.isOperationDate(date);
+        if (!isOperationDate) {
+            return;
+        }
```
```suggestion
        if (!CampusManager.isOperationDate(date)) {
            return;
        }
```

요게 가독에 더 좋겠네요. 메서드내 변수들도 메서드내에서 따로 쓰이지않는다면 굳이 추출할 필요는 없는거같아요

JunHoPark93 commented on src/main/java/attendance/domain/AttendanceHistory.java:
```diff
@@ -1,49 +1,82 @@
+        }
+        Optional<Attendance> attendance = findAttendance(date);
+        if (attendance.isEmpty()) {
+            statistics.put(ABSENCE, statistics.get(ABSENCE) + 1);
+            return;
         }
-        AttendanceHistory that = (AttendanceHistory) o;
-        return Objects.equals(attendanceDateTime.toLocalDate(), that.attendanceDateTime.toLocalDate());
+        AttendanceStatus status = attendance.get().getStatus();
+        statistics.put(status, statistics.get(status) + 1);
```
취향차이일수도있는데 optional의 orElse, orElseGet등을 사용해보시는것도 고려해보시면 좋을거같아요

JunHoPark93 reviewed (approved):
안녕하세요 레오~
리뷰반영 잘 해주셨네요.
질문주신부분은 코멘트 다시 남겨두었는데요,
확인해보시고 해결되지 않았다면 dm부탁드릴게요~
고생많으셨고 다음 미션도 화이팅입니다!
